transform_aes.c

```diff
@@ -235,60 +210,99 @@ static int transop_decode_aes( n2n_trans_op_t * arg,
+                            len );
                 } else
-                    traceEvent(TRACE_WARNING, "UDP payload decryption failed.");
+                    traceEvent( TRACE_WARNING, "UDP payload decryption failed." );
             } else {
-                traceEvent(TRACE_WARNING, "Encrypted length %d is not a multiple of AES_BLOCK_SIZE (%d)", len, AES_BLOCK_SIZE);
+                traceEvent( TRACE_WARNING, "Encrypted length %d is not a multiple of AES_BLOCK_SIZE (%d)", len, AES_BLOCK_SIZE );
                 len = 0;
             }
         } else
-            traceEvent(TRACE_ERROR, "decode_aes unsupported aes version %u.", aes_enc_ver);
+            traceEvent( TRACE_ERROR, "decode_aes unsupported aes version %u.", aes_enc_ver );
     } else
-        traceEvent(TRACE_ERROR, "decode_aes inbuf wrong size (%ul) to decrypt.", in_len);
+        traceEvent( TRACE_ERROR, "decode_aes inbuf wrong size (%ul) to decrypt.", in_len );
 
     return len;
 }
 
 static int setup_aes_key(transop_aes_t *priv, const uint8_t *key, ssize_t key_size) {
-    size_t aes_keysize_bytes;
-    size_t aes_keysize_bits;
-    struct sha512_keybuf keybuf;
+    size_t aes_key_size_bytes;
+    size_t aes_key_size_bits;
 
     /* Clear out any old possibly longer key matter. */
-    memset( &(priv->enc_key), 0, sizeof(priv->enc_key));
-    memset( &(priv->dec_key), 0, sizeof(priv->dec_key));
-    memset( &(priv->iv_enc_key), 0, sizeof(priv->iv_enc_key));
-    memset( &(priv->iv_ext_val), 0, sizeof(priv->iv_ext_val));
-
-    /* We still use aes_best_keysize (even not necessary since we hash the key
-     * into the 256bits enc_dec_key) to let the users choose the degree of encryption.
-     * Long keys will pick AES192 or AES256 with more robust but expensive encryption.
+    memset( &(priv->enc_key), 0, sizeof(priv->enc_key) );
+    memset( &(priv->dec_key), 0, sizeof(priv->dec_key) );
+    memset( &(priv->iv_enc_key), 0, sizeof(priv->iv_enc_key) );
+    memset( &(priv->iv_pad_val), 0, sizeof(priv->iv_pad_val) );
+
+    /* Let the user choose the degree of encryption:
+     * Long input keys will pick AES192 or AES256 with more robust but expensive encryption.
+     *
+     * The input key always gets hashed to make a more unpredictable use of the key space and
+     * also to derive some additional material (key for IV encrpytion, IV padding).
+     *
+     * The following scheme for key setup was discussed on github: 
+     * https://github.com/ntop/n2n/issues/101
      */
-    aes_keysize_bytes = aes_best_keysize(key_size);
-    aes_keysize_bits = 8 * aes_keysize_bytes;
 
-    /* Hash the main key to generate subkeys */
-    SHA512(key, key_size, (u_char*)&keybuf);
+    /* create a working buffer of maximal occuring hashes size and generate
+     * the hashes for the aes key material, key_mat_buf_lengh indicates the
+     * actual "filling level" of the buffer
+     */
+    uint8_t key_mat_buf[SHA512_DIGEST_LENGTH + SHA256_DIGEST_LENGTH];
```

Comment:
Move declarations at the start of the function otherwise some compilers will argue

Reply:
Okay, applied.